Setting DAG Task dependencies through `next` is too restrictive

**Is your feature request related to a problem? Please describe.**
Given the following Tasks:

```py
with DAG(name="dag"):
    A = Task(...)
    B = Task(...)
    C = Task(...)
```

I want to describe the dependencies:
* Run B if A Failed or Errored
* Run C if A Succeeded or Skipped

Which have to be described using strings in `depends`:

```py
B.depends = "A.Failed || A.Errored"
C.depends = "A.Succeeded"
```

Because trying to use `next` like so (which is valid syntactically):

```py
        A.next(B, on=TaskResult.failed)
        A.next(B, operator=Operator.or_, on=TaskResult.errored)
        A.next(C, on=TaskResult.succeeded)
```

results in:

```
>           raise ValueError(f"{self.name} already in {other.name}'s depends: {other.depends}")
E           ValueError: A already in B's depends: A.Failed

src/hera/workflows/task.py:84: ValueError
```

**Describe the solution you'd like**
A better way to describe more complex conditions

**Describe alternatives you've considered**
Use `depends` directly

**Additional context**
Add any other context or screenshots about the feature request here.

--- a/docs/examples/workflows/dags/dag_configurable_rshift.md
+++ b/docs/examples/workflows/dags/dag_configurable_rshift.md
@@ -1,0 +1,121 @@
+# Dag Configurable Rshift
+
+
+
+This example shows how to override the defaults for the "next" function to configure the rshift (`>>`) behaviour.
+
+This is useful if you want `A >> B` to mean "run B _only if_ A succeeded", otherwise the
+[default depends logic](https://argo-workflows.readthedocs.io/en/latest/enhanced-depends-logic/) means `A >> B` is
+equivalent to "B depends on `A.Succeeded || A.Skipped || A.Daemoned`".
+
+By setting the values in `Task.set_next_defaults`, we can configure the rshift behaviour to use a different operator
+and TaskResult. Then, the following
+
+```py
+with Task.set_next_defaults(operator=Operator.or_, on=TaskResult.succeeded):
+    A >> [B, C] >> D
+```
+
+is equivalent to
+
+```py
+A.next(B, on=TaskResult.succeeded)
+A.next(C, on=TaskResult.succeeded)
+B.next(D, on=TaskResult.succeeded)
+C.next(D, operator=Operator.or_, on=TaskResult.succeeded)
+```
+
+> Note the `Operator.or_` for D's `depends` is set when calling `C.next` which can also be confusing! This is because we
+> use `next` to describe the forward relationships, while the Argo field is `depends` which describes the backward
+> relationships.
+
+Or, described using the backward relationship of `depends` (which only accepts strings):
+```py
+B.depends = "A.Succeeded"
+C.depends = "A.Succeeded"
+D.depends = "B.Succeeded || C.Succeeded"
+```
+
+
+=== "Hera"
+
+    ```python linenums="1"
+    from hera.workflows import DAG, Task, TaskResult, Workflow, script
+    from hera.workflows.operator import Operator
+
+
+    @script(image="python:3.12")
+    def echo(message):
+        print(message)
+
+
+    with Workflow(generate_name="dag-configurable-rshift-", entrypoint="diamond") as w:
+        with DAG(name="diamond"):
+            A = echo(name="A", arguments={"message": "A"})
+            B = echo(name="B", arguments={"message": "B"})
+            C = echo(name="C", arguments={"message": "C"})
+            D = echo(name="D", arguments={"message": "D"})
+
+            with Task.set_next_defaults(operator=Operator.or_, on=TaskResult.succeeded):
+                A >> [B, C] >> D
+    ```
+
+=== "YAML"
+
+    ```yaml linenums="1"
+    apiVersion: argoproj.io/v1alpha1
+    kind: Workflow
+    metadata:
+      generateName: dag-configurable-rshift-
+    spec:
+      entrypoint: diamond
+      templates:
+      - name: diamond
+        dag:
+          tasks:
+          - name: A
+            template: echo
+            arguments:
+              parameters:
+              - name: message
+                value: A
+          - name: B
+            depends: A.Succeeded
+            template: echo
+            arguments:
+              parameters:
+              - name: message
+                value: B
+          - name: C
+            depends: A.Succeeded
+            template: echo
+            arguments:
+              parameters:
+              - name: message
+                value: C
+          - name: D
+            depends: B.Succeeded || C.Succeeded
+            template: echo
+            arguments:
+              parameters:
+              - name: message
+                value: D
+      - name: echo
+        inputs:
+          parameters:
+          - name: message
+        script:
+          image: python:3.12
+          source: |-
+            import os
+            import sys
+            sys.path.append(os.getcwd())
+            import json
+            try: message = json.loads(r'''{{inputs.parameters.message}}''')
+            except: message = r'''{{inputs.parameters.message}}'''
+
+            print(message)
+          command:
+          - python
+    ```
+

--- a/docs/examples/workflows/dags/dag_configurable_rshift.md
+++ b/docs/examples/workflows/dags/dag_configurable_rshift.md
@@ -36,6 +36,8 @@ C.depends = "A.Succeeded"
 D.depends = "B.Succeeded || C.Succeeded"
 ```
 
+> `set_next_defaults` also accepts lists or multiple values using the `|` operator!
+
 
 === "Hera"
 

--- a/docs/walk-through/dags.md
+++ b/docs/walk-through/dags.md
@@ -271,4 +271,13 @@ C.depends = "A.Succeeded"
 D.depends = "B.Succeeded || C.Succeeded"
 ```
 
+The `set_next_defaults` function also accepts lists or values applying the `|` operator for the `on` value, meaning you can also specify conditions like `B.depends = "A.Succeeded || A.Daemoned"`, without affecting the `operator` used. E.g:
+
+```py
+with Task.set_next_defaults(operator=Operator.and_, on=TaskResult.succeeded | TaskResult.skipped):
+    [task_a, task_b] >> task_c
+
+assert task_c.depends == "(task-a.Skipped || task-a.Succeeded) && (task-b.Skipped || task-b.Succeeded)"
+```
+
 See the [DAG Configurable rshift example](../examples/workflows/dags/dag_configurable_rshift.md) for the full code!

--- a/docs/walk-through/dags.md
+++ b/docs/walk-through/dags.md
@@ -233,3 +233,42 @@ A   B
 | X |
 C   D
 ```
+
+## Configuring the Default "Next" Behaviour for `>>`
+
+Hera v5.24 added the `Task.set_next_defaults` function, allowing you to set the default `operator` and `on` values
+within a scoped context, which by extension allows you to configure the rshift (`>>`) behaviour.
+
+This is useful if you want `A >> B` to mean "run B _only if_ A succeeded", otherwise the
+[default depends logic](https://argo-workflows.readthedocs.io/en/latest/enhanced-depends-logic/) means `A >> B` is
+equivalent to "B depends on `A.Succeeded || A.Skipped || A.Daemoned`".
+
+By setting the values in `Task.set_next_defaults`, we can configure the rshift behaviour to use a different operator
+and TaskResult. Then, the following
+
+```py
+with Task.set_next_defaults(operator=Operator.or_, on=TaskResult.succeeded):
+    A >> [B, C] >> D
+```
+
+is equivalent to
+
+```py
+A.next(B, on=TaskResult.succeeded)
+A.next(C, on=TaskResult.succeeded)
+B.next(D, on=TaskResult.succeeded)
+C.next(D, operator=Operator.or_, on=TaskResult.succeeded)
+```
+
+> Note the `Operator.or_` for D's `depends` is set when calling `C.next` which can also be confusing! This is because we
+> use `next` to describe the forward relationships, while the Argo field is `depends` which describes the backward
+> relationships.
+
+Or, described using the backward relationship of `depends` (which only accepts strings):
+```py
+B.depends = "A.Succeeded"
+C.depends = "A.Succeeded"
+D.depends = "B.Succeeded || C.Succeeded"
+```
+
+See the [DAG Configurable rshift example](../examples/workflows/dags/dag_configurable_rshift.md) for the full code!

--- a/examples/workflows/dags/dag-configurable-rshift.yaml
+++ b/examples/workflows/dags/dag-configurable-rshift.yaml
@@ -1,0 +1,54 @@
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: dag-configurable-rshift-
+spec:
+  entrypoint: diamond
+  templates:
+  - name: diamond
+    dag:
+      tasks:
+      - name: A
+        template: echo
+        arguments:
+          parameters:
+          - name: message
+            value: A
+      - name: B
+        depends: A.Succeeded
+        template: echo
+        arguments:
+          parameters:
+          - name: message
+            value: B
+      - name: C
+        depends: A.Succeeded
+        template: echo
+        arguments:
+          parameters:
+          - name: message
+            value: C
+      - name: D
+        depends: B.Succeeded || C.Succeeded
+        template: echo
+        arguments:
+          parameters:
+          - name: message
+            value: D
+  - name: echo
+    inputs:
+      parameters:
+      - name: message
+    script:
+      image: python:3.12
+      source: |-
+        import os
+        import sys
+        sys.path.append(os.getcwd())
+        import json
+        try: message = json.loads(r'''{{inputs.parameters.message}}''')
+        except: message = r'''{{inputs.parameters.message}}'''
+
+        print(message)
+      command:
+      - python

--- a/examples/workflows/dags/dag_configurable_rshift.py
+++ b/examples/workflows/dags/dag_configurable_rshift.py
@@ -1,0 +1,53 @@
+"""This example shows how to override the defaults for the "next" function to configure the rshift (`>>`) behaviour.
+
+This is useful if you want `A >> B` to mean "run B _only if_ A succeeded", otherwise the
+[default depends logic](https://argo-workflows.readthedocs.io/en/latest/enhanced-depends-logic/) means `A >> B` is
+equivalent to "B depends on `A.Succeeded || A.Skipped || A.Daemoned`".
+
+By setting the values in `Task.set_next_defaults`, we can configure the rshift behaviour to use a different operator
+and TaskResult. Then, the following
+
+```py
+with Task.set_next_defaults(operator=Operator.or_, on=TaskResult.succeeded):
+    A >> [B, C] >> D
+```
+
+is equivalent to
+
+```py
+A.next(B, on=TaskResult.succeeded)
+A.next(C, on=TaskResult.succeeded)
+B.next(D, on=TaskResult.succeeded)
+C.next(D, operator=Operator.or_, on=TaskResult.succeeded)
+```
+
+> Note the `Operator.or_` for D's `depends` is set when calling `C.next` which can also be confusing! This is because we
+> use `next` to describe the forward relationships, while the Argo field is `depends` which describes the backward
+> relationships.
+
+Or, described using the backward relationship of `depends` (which only accepts strings):
+```py
+B.depends = "A.Succeeded"
+C.depends = "A.Succeeded"
+D.depends = "B.Succeeded || C.Succeeded"
+```
+"""
+
+from hera.workflows import DAG, Task, TaskResult, Workflow, script
+from hera.workflows.operator import Operator
+
+
+@script(image="python:3.12")
+def echo(message):
+    print(message)
+
+
+with Workflow(generate_name="dag-configurable-rshift-", entrypoint="diamond") as w:
+    with DAG(name="diamond"):
+        A = echo(name="A", arguments={"message": "A"})
+        B = echo(name="B", arguments={"message": "B"})
+        C = echo(name="C", arguments={"message": "C"})
+        D = echo(name="D", arguments={"message": "D"})
+
+        with Task.set_next_defaults(operator=Operator.or_, on=TaskResult.succeeded):
+            A >> [B, C] >> D

--- a/examples/workflows/dags/dag_configurable_rshift.py
+++ b/examples/workflows/dags/dag_configurable_rshift.py
@@ -31,6 +31,8 @@ B.depends = "A.Succeeded"
 C.depends = "A.Succeeded"
 D.depends = "B.Succeeded || C.Succeeded"
 ```
+
+> `set_next_defaults` also accepts lists or multiple values using the `|` operator!
 """
 
 from hera.workflows import DAG, Task, TaskResult, Workflow, script

--- a/src/hera/workflows/task.py
+++ b/src/hera/workflows/task.py
@@ -130,7 +130,7 @@ class Task(
         on: OnType = None,
     ) -> Task:
         """Set self as a dependency of `other`."""
-        operator = operator if operator is not None else self.__class__._default_next_operator
+        operator = operator or self.__class__._default_next_operator
         on_list = _normalise_on(on, self.__class__._default_next_on)
 
         # Build condition string:

--- a/src/hera/workflows/task.py
+++ b/src/hera/workflows/task.py
@@ -8,7 +8,7 @@ from __future__ import annotations
 
 from contextlib import contextmanager
 from enum import Enum
-from typing import ClassVar, List, Optional, Union
+from typing import ClassVar, Iterable, Iterator, List, Optional, Set, Union
 
 from hera.workflows._mixins import (
     ArgumentsMixin,
@@ -43,6 +43,47 @@ class TaskResult(Enum):
     any_succeeded = "AnySucceeded"
     all_failed = "AllFailed"
 
+    def __or__(self, other: TaskResult) -> Iterable[TaskResult]:
+        """Create an "or" condition over multiple TaskResults."""
+        return _TaskResultGroup({self, other})
+
+
+class _TaskResultGroup(Iterable[TaskResult]):
+    """Private, implementation detail: an iterable of TaskResult with | support."""
+
+    __slots__ = ("_results",)
+
+    def __init__(self, results: Iterable[TaskResult]):
+        # use set for idempotence
+        self._results: Set[TaskResult] = set(results)
+
+    def __or__(self, other: Union[Iterable[TaskResult], TaskResult]):
+        if isinstance(other, TaskResult):
+            return _TaskResultGroup(self._results | {other})
+        # If user chained with another group or any iterable, merge
+        return _TaskResultGroup(self._results | set(other))
+
+    def __iter__(self) -> Iterator[TaskResult]:
+        return iter(self._results)
+
+
+def _normalize_on(
+    on: Union[TaskResult, Iterable[TaskResult], None],
+    default: Optional[List[TaskResult]],
+) -> Optional[List[TaskResult]]:
+    """Turn `on` into a list[TaskResult] or None.
+
+    Accepts:
+      - None -> return default
+      - TaskResult -> [TaskResult]
+      - Iterable[TaskResult] (including private group, sets, tuples, generators) -> list(...)
+    """
+    if on is None:
+        return default
+    if isinstance(on, TaskResult):
+        return [on]
+    return list(on)
+
 
 class Task(
     TemplateInvocatorSubNodeMixin,
@@ -57,7 +98,7 @@ class Task(
     depends: Optional[str] = None
 
     _default_next_operator: ClassVar[Operator] = Operator.and_
-    _default_next_on: ClassVar[Optional[TaskResult]] = None
+    _default_next_on: ClassVar[Optional[List[TaskResult]]] = None
 
     def _get_dependency_tasks(self) -> List[str]:
         if self.depends is None:
@@ -75,38 +116,58 @@ class Task(
     def _subtype(self) -> str:
         return "tasks"
 
-    def next(self, other: Task, operator: Optional[Operator] = None, on: Optional[TaskResult] = None) -> Task:
+    def next(
+        self,
+        other: Task,
+        operator: Optional[Operator] = None,
+        on: Union[TaskResult, Iterable[TaskResult], None] = None,
+    ) -> Task:
         """Set self as a dependency of `other`."""
-        assert issubclass(other.__class__, Task)
+        operator = operator if operator is not None else self.__class__._default_next_operator
+        on_list = _normalize_on(on, self.__class__._default_next_on)
 
-        operator = operator or self.__class__._default_next_operator
-        on = on or self.__class__._default_next_on
-
-        condition = f".{on.value}" if on else ""
+        # Build condition string:
+        # - If multiple on-conditions: OR them and wrap in parens
+        # - If single on-condition: A.succeeded
+        # - If none: just "A"
+        if on_list and len(on_list) > 1:
+            condition_str = " || ".join(f"{self.name}.{c.value}" for c in on_list)
+            condition_str = f"({condition_str})"
+        elif on_list and len(on_list) == 1:
+            condition_str = f"{self.name}.{on_list[0].value}"
+        else:
+            condition_str = self.name
 
         if other.depends is None:
             # First dependency
-            other.depends = self.name + condition
+            other.depends = condition_str
         elif self.name in other._get_dependency_tasks():
             raise ValueError(f"{self.name} already in {other.name}'s depends: {other.depends}")
         else:
             # Add follow-up dependency
-            other.depends += f" {operator} {self.name + condition}"
+            other.depends += f" {operator} {condition_str}"
+
         return other
 
     @classmethod
     @contextmanager
-    def set_next_defaults(cls, operator: Operator = Operator.and_, on: Optional[TaskResult] = None):
-        """Temporarily set the defaults used by Task.next."""
-        old_operator_default = cls._default_next_operator
-        old_on_default = cls._default_next_on
+    def set_next_defaults(
+        cls,
+        operator: Operator = Operator.and_,
+        on: Union[TaskResult, Iterable[TaskResult], None] = None,
+    ):
+        """Temporarily override the default operator and on."""
+        on_list = _normalize_on(on, None)
+
+        old_operator = cls._default_next_operator
+        old_on = cls._default_next_on
         cls._default_next_operator = operator
-        cls._default_next_on = on
+        cls._default_next_on = on_list
         try:
             yield
         finally:
-            cls._default_next_operator = old_operator_default
-            cls._default_next_on = old_on_default
+            cls._default_next_operator = old_operator
+            cls._default_next_on = old_on
 
     def __rrshift__(self, other: List[Union[Task, str]]) -> Task:
         """Set `other` as a dependency self."""
@@ -138,7 +199,7 @@ class Task(
         raise ValueError(f"Unknown type {type(other)} provided to `__rshift__`")
 
     def __or__(self, other: Union[Task, str]) -> str:
-        """Adds a condition of."""
+        """Return a condition of `self || other`."""
         if isinstance(other, Task):
             return f"({self.name} || {other.name})"
         assert isinstance(other, str), f"Unknown type {type(other)} specified using `|` operator"

--- a/src/hera/workflows/task.py
+++ b/src/hera/workflows/task.py
@@ -6,8 +6,9 @@ for more on using Tasks within a DAG.
 
 from __future__ import annotations
 
+from contextlib import contextmanager
 from enum import Enum
-from typing import List, Optional, Union
+from typing import ClassVar, List, Optional, Union
 
 from hera.workflows._mixins import (
     ArgumentsMixin,
@@ -55,6 +56,9 @@ class Task(
     dependencies: Optional[List[str]] = None
     depends: Optional[str] = None
 
+    _default_next_operator: ClassVar[Operator] = Operator.and_
+    _default_next_on: ClassVar[Optional[TaskResult]] = None
+
     def _get_dependency_tasks(self) -> List[str]:
         if self.depends is None:
             return []
@@ -71,9 +75,12 @@ class Task(
     def _subtype(self) -> str:
         return "tasks"
 
-    def next(self, other: Task, operator: Operator = Operator.and_, on: Optional[TaskResult] = None) -> Task:
+    def next(self, other: Task, operator: Optional[Operator] = None, on: Optional[TaskResult] = None) -> Task:
         """Set self as a dependency of `other`."""
         assert issubclass(other.__class__, Task)
+
+        operator = operator or self.__class__._default_next_operator
+        on = on or self.__class__._default_next_on
 
         condition = f".{on.value}" if on else ""
 
@@ -86,6 +93,20 @@ class Task(
             # Add follow-up dependency
             other.depends += f" {operator} {self.name + condition}"
         return other
+
+    @classmethod
+    @contextmanager
+    def set_next_defaults(cls, operator: Operator = Operator.and_, on: Optional[TaskResult] = None):
+        """Temporarily set the defaults used by Task.next."""
+        old_operator_default = cls._default_next_operator
+        old_on_default = cls._default_next_on
+        cls._default_next_operator = operator
+        cls._default_next_on = on
+        try:
+            yield
+        finally:
+            cls._default_next_operator = old_operator_default
+            cls._default_next_on = old_on_default
 
     def __rrshift__(self, other: List[Union[Task, str]]) -> Task:
         """Set `other` as a dependency self."""

--- a/src/hera/workflows/task.py
+++ b/src/hera/workflows/task.py
@@ -8,7 +8,7 @@ from __future__ import annotations
 
 from contextlib import contextmanager
 from enum import Enum
-from typing import ClassVar, Iterable, Iterator, List, Optional, Set, Union
+from typing import ClassVar, Iterable, Iterator, List, Optional, Union
 
 from hera.workflows._mixins import (
     ArgumentsMixin,
@@ -45,44 +45,51 @@ class TaskResult(Enum):
 
     def __or__(self, other: TaskResult) -> Iterable[TaskResult]:
         """Create an "or" condition over multiple TaskResults."""
-        return _TaskResultGroup({self, other})
+        return _TaskResultGroup([self, other])
 
 
 class _TaskResultGroup(Iterable[TaskResult]):
-    """Private, implementation detail: an iterable of TaskResult with | support."""
+    """Private, implementation detail: an iterable of TaskResult with | support that maintains order and deduplicates."""
 
     __slots__ = ("_results",)
 
     def __init__(self, results: Iterable[TaskResult]):
-        # use set for idempotence
-        self._results: Set[TaskResult] = set(results)
+        self._results: List[TaskResult] = []
+        for r in results:
+            if r not in self._results:
+                self._results.append(r)
 
     def __or__(self, other: Union[Iterable[TaskResult], TaskResult]):
         if isinstance(other, TaskResult):
-            return _TaskResultGroup(self._results | {other})
-        # If user chained with another group or any iterable, merge
-        return _TaskResultGroup(self._results | set(other))
+            return _TaskResultGroup(self._results + [other])
+        return _TaskResultGroup(self._results + list(other))
 
     def __iter__(self) -> Iterator[TaskResult]:
         return iter(self._results)
 
 
-def _normalize_on(
-    on: Union[TaskResult, Iterable[TaskResult], None],
-    default: Optional[List[TaskResult]],
+OnType = Optional[Union[TaskResult, Iterable[TaskResult]]]
+
+
+def _normalise_on(
+    on: OnType,
+    default: OnType = None,
 ) -> Optional[List[TaskResult]]:
     """Turn `on` into a list[TaskResult] or None.
 
     Accepts:
-      - None -> return default
+      - None -> return normalised default (which may also be None)
       - TaskResult -> [TaskResult]
-      - Iterable[TaskResult] (including private group, sets, tuples, generators) -> list(...)
+      - Iterable[TaskResult] -> list(...)
     """
     if on is None:
-        return default
+        if default is None:
+            return None
+
+        return _normalise_on(default)
     if isinstance(on, TaskResult):
         return [on]
-    return list(on)
+    return list(_TaskResultGroup(on))
 
 
 class Task(
@@ -98,7 +105,7 @@ class Task(
     depends: Optional[str] = None
 
     _default_next_operator: ClassVar[Operator] = Operator.and_
-    _default_next_on: ClassVar[Optional[List[TaskResult]]] = None
+    _default_next_on: ClassVar[OnType] = None
 
     def _get_dependency_tasks(self) -> List[str]:
         if self.depends is None:
@@ -120,11 +127,11 @@ class Task(
         self,
         other: Task,
         operator: Optional[Operator] = None,
-        on: Union[TaskResult, Iterable[TaskResult], None] = None,
+        on: OnType = None,
     ) -> Task:
         """Set self as a dependency of `other`."""
         operator = operator if operator is not None else self.__class__._default_next_operator
-        on_list = _normalize_on(on, self.__class__._default_next_on)
+        on_list = _normalise_on(on, self.__class__._default_next_on)
 
         # Build condition string:
         # - If multiple on-conditions: OR them and wrap in parens
@@ -157,7 +164,7 @@ class Task(
         on: Union[TaskResult, Iterable[TaskResult], None] = None,
     ):
         """Temporarily override the default operator and on."""
-        on_list = _normalize_on(on, None)
+        on_list = _normalise_on(on)
 
         old_operator = cls._default_next_operator
         old_on = cls._default_next_on

--- a/src/hera/workflows/task.py
+++ b/src/hera/workflows/task.py
@@ -160,15 +160,15 @@ class Task(
     @contextmanager
     def set_next_defaults(
         cls,
-        operator: Operator = Operator.and_,
+        operator: Optional[Operator] = None,
         on: Union[TaskResult, Iterable[TaskResult], None] = None,
     ):
-        """Temporarily override the default operator and on."""
+        """Temporarily modify the default behaviour of `next` and `>>`."""
         on_list = _normalise_on(on)
 
         old_operator = cls._default_next_operator
         old_on = cls._default_next_on
-        cls._default_next_operator = operator
+        cls._default_next_operator = operator or cls._default_next_operator
         cls._default_next_on = on_list
         try:
             yield

--- a/src/hera/workflows/task.py
+++ b/src/hera/workflows/task.py
@@ -7,8 +7,9 @@ for more on using Tasks within a DAG.
 from __future__ import annotations
 
 from contextlib import contextmanager
+from contextvars import ContextVar
 from enum import Enum
-from typing import ClassVar, Iterable, Iterator, List, Optional, Union
+from typing import Iterable, Iterator, List, Optional, Union
 
 from hera.workflows._mixins import (
     ArgumentsMixin,
@@ -89,6 +90,10 @@ def _normalise_on(
     return list(_TaskResultGroup(on))
 
 
+_default_next_operator: ContextVar[Operator] = ContextVar("_default_next_operator", default=Operator.and_)
+_default_next_on: ContextVar[Optional[List[TaskResult]]] = ContextVar("_default_on_operator", default=None)
+
+
 class Task(
     TemplateInvocatorSubNodeMixin,
     ArgumentsMixin,
@@ -100,9 +105,6 @@ class Task(
 
     dependencies: Optional[List[str]] = None
     depends: Optional[str] = None
-
-    _default_next_operator: ClassVar[Operator] = Operator.and_
-    _default_next_on: ClassVar[Optional[List[TaskResult]]] = None
 
     def _get_dependency_tasks(self) -> List[str]:
         if self.depends is None:
@@ -127,8 +129,8 @@ class Task(
         on: OnType = None,
     ) -> Task:
         """Set self as a dependency of `other`."""
-        operator = operator or self.__class__._default_next_operator
-        on_list = _normalise_on(on, self.__class__._default_next_on)
+        operator = operator or _default_next_operator.get()
+        on_list = _normalise_on(on, _default_next_on.get())
 
         # Build condition string:
         # - If multiple on-conditions: OR them and wrap in parens
@@ -163,15 +165,15 @@ class Task(
         """Temporarily modify the default behaviour of `next` and `>>`."""
         on_list = _normalise_on(on)
 
-        old_operator = cls._default_next_operator
-        old_on = cls._default_next_on
-        cls._default_next_operator = operator or cls._default_next_operator
-        cls._default_next_on = on_list
+        old_operator = _default_next_operator.get()
+        old_on = _default_next_on.get()
+        _default_next_operator.set(operator or _default_next_operator.get())
+        _default_next_on.set(on_list)
         try:
             yield
         finally:
-            cls._default_next_operator = old_operator
-            cls._default_next_on = old_on
+            _default_next_operator.set(old_operator)
+            _default_next_on.set(old_on)
 
     def __rrshift__(self, other: List[Union[Task, str]]) -> Task:
         """Set `other` as a dependency self."""

--- a/src/hera/workflows/task.py
+++ b/src/hera/workflows/task.py
@@ -73,20 +73,17 @@ OnType = Optional[Union[TaskResult, Iterable[TaskResult]]]
 
 def _normalise_on(
     on: OnType,
-    default: OnType = None,
+    default: Optional[List[TaskResult]] = None,
 ) -> Optional[List[TaskResult]]:
     """Turn `on` into a list[TaskResult] or None.
 
     Accepts:
-      - None -> return normalised default (which may also be None)
+      - None -> return default (which may also be None)
       - TaskResult -> [TaskResult]
       - Iterable[TaskResult] -> list(...)
     """
     if on is None:
-        if default is None:
-            return None
-
-        return _normalise_on(default)
+        return default
     if isinstance(on, TaskResult):
         return [on]
     return list(_TaskResultGroup(on))
@@ -105,7 +102,7 @@ class Task(
     depends: Optional[str] = None
 
     _default_next_operator: ClassVar[Operator] = Operator.and_
-    _default_next_on: ClassVar[OnType] = None
+    _default_next_on: ClassVar[Optional[List[TaskResult]]] = None
 
     def _get_dependency_tasks(self) -> List[str]:
         if self.depends is None:

--- a/src/hera/workflows/task.py
+++ b/src/hera/workflows/task.py
@@ -43,7 +43,7 @@ class TaskResult(Enum):
     any_succeeded = "AnySucceeded"
     all_failed = "AllFailed"
 
-    def __or__(self, other: TaskResult) -> Iterable[TaskResult]:
+    def __or__(self, other: TaskResult) -> _TaskResultGroup:
         """Create an "or" condition over multiple TaskResults."""
         return _TaskResultGroup([self, other])
 
@@ -59,7 +59,7 @@ class _TaskResultGroup(Iterable[TaskResult]):
             if r not in self._results:
                 self._results.append(r)
 
-    def __or__(self, other: Union[Iterable[TaskResult], TaskResult]):
+    def __or__(self, other: Union[Iterable[TaskResult], TaskResult]) -> _TaskResultGroup:
         if isinstance(other, TaskResult):
             return _TaskResultGroup(self._results + [other])
         return _TaskResultGroup(self._results + list(other))

--- a/tests/test_unit/test_task.py
+++ b/tests/test_unit/test_task.py
@@ -1,0 +1,43 @@
+from hera.workflows.operator import Operator
+from hera.workflows.task import Task, TaskResult
+
+
+def test_task_rshift():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+
+    task_a >> task_b
+
+    assert task_b.depends == "task-a"
+
+
+def test_override_task_next_success_only():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+
+    with Task.set_next_defaults(on=TaskResult.succeeded):
+        task_a >> task_b
+
+    assert task_b.depends == "task-a.Succeeded"
+
+
+def test_override_task_next_operator_or():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+    task_c = Task(name="task-c", template="dummy")
+
+    with Task.set_next_defaults(operator=Operator.or_):
+        [task_a, task_b] >> task_c
+
+    assert task_c.depends == "task-a || task-b"
+
+
+def test_override_task_next_operator_or_success_only():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+    task_c = Task(name="task-c", template="dummy")
+
+    with Task.set_next_defaults(operator=Operator.or_, on=TaskResult.succeeded):
+        [task_a, task_b] >> task_c
+
+    assert task_c.depends == "task-a.Succeeded || task-b.Succeeded"

--- a/tests/test_unit/test_task.py
+++ b/tests/test_unit/test_task.py
@@ -11,6 +11,16 @@ def test_task_rshift():
     assert task_b.depends == "task-a"
 
 
+def test_task_or_for_depends():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+    task_c = Task(name="task-c", template="dummy")
+
+    task_c.depends = task_a | task_b
+
+    assert task_c.depends == "(task-a || task-b)"
+
+
 def test_override_task_next_success_only():
     task_a = Task(name="task-a", template="dummy")
     task_b = Task(name="task-b", template="dummy")
@@ -21,14 +31,56 @@ def test_override_task_next_success_only():
     assert task_b.depends == "task-a.Succeeded"
 
 
-def test_task_rshift_or():
+def test_override_task_next_success_or_skipped():
     task_a = Task(name="task-a", template="dummy")
     task_b = Task(name="task-b", template="dummy")
-    task_c = Task(name="task-c", template="dummy")
 
-    task_c.depends = task_a | task_b
+    with Task.set_next_defaults(on=TaskResult.succeeded | TaskResult.skipped):
+        task_a >> task_b
 
-    assert task_c.depends == "(task-a || task-b)"
+    assert task_b.depends == "(task-a.Succeeded || task-a.Skipped)"
+
+
+def test_override_task_next_success_or_skipped_list():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+
+    with Task.set_next_defaults(on=[TaskResult.succeeded, TaskResult.skipped]):
+        task_a >> task_b
+
+    assert task_b.depends == "(task-a.Succeeded || task-a.Skipped)"
+
+
+def test_override_task_next_ignore_duplicate_task_results():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+
+    with Task.set_next_defaults(on=TaskResult.succeeded | TaskResult.succeeded):
+        task_a >> task_b
+
+    assert task_b.depends == "task-a.Succeeded"
+
+
+def test_override_task_next_ignore_duplicate_task_results_list():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+
+    with Task.set_next_defaults(on=[TaskResult.succeeded, TaskResult.succeeded]):
+        task_a >> task_b
+
+    assert task_b.depends == "task-a.Succeeded"
+
+
+def test_override_task_next_ignore_multiple_duplicate_task_results():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+
+    with Task.set_next_defaults(
+        on=TaskResult.succeeded | TaskResult.skipped | TaskResult.succeeded | TaskResult.skipped
+    ):
+        task_a >> task_b
+
+    assert task_b.depends == "(task-a.Succeeded || task-a.Skipped)"
 
 
 def test_override_task_next_operator_or():
@@ -72,7 +124,7 @@ def test_override_task_next_on_success_or_skipped_or_operator():
     with Task.set_next_defaults(on=TaskResult.succeeded | TaskResult.skipped):
         [task_a, task_b] >> task_c
 
-    assert task_c.depends == "(task-a.Skipped || task-a.Succeeded) && (task-b.Skipped || task-b.Succeeded)"
+    assert task_c.depends == "(task-a.Succeeded || task-a.Skipped) && (task-b.Succeeded || task-b.Skipped)"
 
 
 def test_override_task_next_or_operator_and_on_success_or_skipped():
@@ -83,4 +135,4 @@ def test_override_task_next_or_operator_and_on_success_or_skipped():
     with Task.set_next_defaults(operator=Operator.or_, on=TaskResult.succeeded | TaskResult.skipped):
         [task_a, task_b] >> task_c
 
-    assert task_c.depends == "(task-a.Skipped || task-a.Succeeded) || (task-b.Skipped || task-b.Succeeded)"
+    assert task_c.depends == "(task-a.Succeeded || task-a.Skipped) || (task-b.Succeeded || task-b.Skipped)"

--- a/tests/test_unit/test_task.py
+++ b/tests/test_unit/test_task.py
@@ -41,6 +41,16 @@ def test_override_task_next_success_or_skipped():
     assert task_b.depends == "(task-a.Succeeded || task-a.Skipped)"
 
 
+def test_override_task_next_success_or_skipped_or_daemoned():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+
+    with Task.set_next_defaults(on=TaskResult.succeeded | TaskResult.skipped | TaskResult.daemoned):
+        task_a >> task_b
+
+    assert task_b.depends == "(task-a.Succeeded || task-a.Skipped || task-a.Daemoned)"
+
+
 def test_override_task_next_success_or_skipped_list():
     task_a = Task(name="task-a", template="dummy")
     task_b = Task(name="task-b", template="dummy")

--- a/tests/test_unit/test_task.py
+++ b/tests/test_unit/test_task.py
@@ -21,6 +21,16 @@ def test_override_task_next_success_only():
     assert task_b.depends == "task-a.Succeeded"
 
 
+def test_task_rshift_or():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+    task_c = Task(name="task-c", template="dummy")
+
+    task_c.depends = task_a | task_b
+
+    assert task_c.depends == "(task-a || task-b)"
+
+
 def test_override_task_next_operator_or():
     task_a = Task(name="task-a", template="dummy")
     task_b = Task(name="task-b", template="dummy")
@@ -41,3 +51,36 @@ def test_override_task_next_operator_or_success_only():
         [task_a, task_b] >> task_c
 
     assert task_c.depends == "task-a.Succeeded || task-b.Succeeded"
+
+
+def test_override_task_next_on_success_or_skipped_list():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+    task_c = Task(name="task-c", template="dummy")
+
+    with Task.set_next_defaults(on=[TaskResult.succeeded, TaskResult.skipped]):
+        [task_a, task_b] >> task_c
+
+    assert task_c.depends == "(task-a.Succeeded || task-a.Skipped) && (task-b.Succeeded || task-b.Skipped)"
+
+
+def test_override_task_next_on_success_or_skipped_or_operator():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+    task_c = Task(name="task-c", template="dummy")
+
+    with Task.set_next_defaults(on=TaskResult.succeeded | TaskResult.skipped):
+        [task_a, task_b] >> task_c
+
+    assert task_c.depends == "(task-a.Skipped || task-a.Succeeded) && (task-b.Skipped || task-b.Succeeded)"
+
+
+def test_override_task_next_or_operator_and_on_success_or_skipped():
+    task_a = Task(name="task-a", template="dummy")
+    task_b = Task(name="task-b", template="dummy")
+    task_c = Task(name="task-c", template="dummy")
+
+    with Task.set_next_defaults(operator=Operator.or_, on=TaskResult.succeeded | TaskResult.skipped):
+        [task_a, task_b] >> task_c
+
+    assert task_c.depends == "(task-a.Skipped || task-a.Succeeded) || (task-b.Skipped || task-b.Succeeded)"

